mail-my-ballot/mail-my-ballot
# Implement AddressForm and StateForm with Styled Modal using custom hooks and Context API

**What has been done**
1. Moved `StateForm` to a Modal to avoid scrolling around in the page
2. Added **custom hook** `useModal` to manage the opening and closing of the modal globally
3. Added `Provider` and `Context` for Styled Modal
4. Added `Modal` component which can be custom styled with `ThemeProvider` for different implementations across the app
5. On success of `AddressForm` API call to `fetchContactAddress` we invoke the `Modal Context` to show the `StateForm` inside the custom modal
6. On submit we call the `Modal Context` again to close the Modal upon successful submit from `states Base` Component

**How to test it**

1. Go to [Deployed Staging FE app link](https://mail-my-ballot.rbprojects7.now.sh/) and [Deployed staging Google Cloud with organisation FE app](https://test-voting-278605.uc.r.appspot.com/)
2. Fill in the Zipcode either manually or select from the dropdown below and fill Zipcode and address
3. After address is filled in, click **Find My Election Official** button

 **Notes**
* We can use Theme context to completely customise the Modal
* The real-benefit of using a Modal for `StateForm` is that its a definitive UX step for the user to understand that some new change is expected instead of scrolling because its a bit more responsive and for cases like mobile and tablet with touch capability, the webpage might not scroll to the desired location
